clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```
@@ -719,7 +718,27 @@ public synchronized void subscribePartition(PubSubTopicPartition topicPartition,
     partitionToPendingConsumerActionCountMap.computeIfAbsent(partitionNumber, x -> new AtomicInteger(0))
         .incrementAndGet();
     pendingSubscriptionActionCount.incrementAndGet();
-    consumerActionsQueue.add(new ConsumerAction(SUBSCRIBE, topicPartition, nextSeqNum(), isHelixTriggeredAction));
+
+    if (recordTransformer != null) {
+      recordTransformerOnRecoveryThreadPool.submit(() -> {
+        try {
+          long startTime = System.nanoTime();
+          recordTransformer.internalOnRecovery(storageEngine, partitionNumber, partitionStateSerializer, compressor);
+          LOGGER.info(
+              "DaVinciRecordTransformer onRecovery took {} ms for replica: {}",
+              LatencyUtils.getElapsedTimeFromNSToMS(startTime),
+              getReplicaId(topicPartition));
+
+          recordTransformer.countDownStartConsumptionLatch();
+          consumerActionsQueue.add(new ConsumerAction(SUBSCRIBE, topicPartition, nextSeqNum(), isHelixTriggeredAction));
+        } catch (Exception e) {
+          LOGGER.error("DaVinciRecordTransformer onRecovery failed for replica: {}", getReplicaId(topicPartition), e);
```

So we don't want to fail loudly??

---

To avoid silent thread terminaton, I created an AtomicBoolean to detect when it does fail. Inside the inifnite while loop in `SIT:run`, we can check if it is true. If so, we fail loudly.

---

Why don't we want to fail fast instead?

---

This piece of code is being executed asychronously in a thread pool. If we threw an exception here, it would not cause the SIT to crash. The AtomicBoolean was one way to detect that failure.

---

For this particular feature, the rocksdb scan is non-blocking and if something bad happens, `IngestionException` will be setup and the periodical run of `checkIngestionProgress` should error out and in `processIngestionException`, you may need to add some logic to cancel all the submitted tasks to `recordTransformerOnRecoveryThreadPool`.

---

If I use `setIngestionException`, `processIngestionException` will just [log the error and unsubscribe the errored partition](https://github.com/linkedin/venice/blob/9c3ad76fcca8aab78dab203c267964af9b0e3e8a/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java#L1464-L1480).

If I use [setLastStoreIngestionException](https://github.com/linkedin/venice/blob/9c3ad76fcca8aab78dab203c267964af9b0e3e8a/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java#L2823-L2825), that will kill the [SIT](https://github.com/linkedin/venice/blob/9c3ad76fcca8aab78dab203c267964af9b0e3e8a/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java#L1485-L1489) in `checkIngestionProgress`. I will be using instead.